nmt.go

```diff
@@ -0,0 +1,142 @@
+package nmt
+
+import (
+	"bytes"
+	"crypto"
+	"errors"
+	"fmt"
+
+	"github.com/liamsi/merkletree"
+)
+
+const (
+	LeafPrefix = 0
+	NodePrefix = 1
+)
+
+var (
+	ErrMismatchedNamespaceSize = errors.New("mismatching namespace sizes")
+)
+
+var _ merkletree.TreeHasher = &namespacedTreeHasher{}
+var _ Nmt = &NamespacedMerkleTree{}
+
+type NamespacedMerkleTree struct {
+	nidLen     int
+	baseHasher crypto.Hash
+	tree       *merkletree.Tree
+}
+
+func (n NamespacedMerkleTree) NamespaceSize() int {
+	return n.nidLen
+}
+
+func (n NamespacedMerkleTree) Push(data NamespacePrefixedData) error {
+	got, want := data.NamespaceSize(), n.NamespaceSize()
+	if got != want {
+		return fmt.Errorf("%w: got: %v, want: %v", ErrMismatchedNamespaceSize, got, want)
+	}
+	// TODO: or should we only push to the actual tree at the end
+	// when we compute the root? The we can push messages bundled by namespaces together: e.g. s.t.
+	// they are lexicographically ordered by namespace IDs
+	// (first transactions,then intermediate state roots, evidence, messages etc)
+	n.tree.Push(data.Bytes())
+	return nil
+}
+
+func (n NamespacedMerkleTree) Root() (minNs, maxNs NamespaceID, root []byte) {
+	tRoot := n.tree.Root()
+	minNs = tRoot[:n.nidLen]
+	maxNs = tRoot[n.nidLen : n.nidLen*2]
+	root = tRoot[n.nidLen*2:]
+	return
+}
+
+//nolint:errcheck
+func (n NamespacedMerkleTree) CompactRoot() []byte {
+	h := n.baseHasher.New()
+	minNs, maxNs, root := n.Root()
+	h.Write(minNs)
+	h.Write(maxNs)
+	h.Write(root)
+	// TODO: should we merkelize these 3 values instead?
+	// Or add in some form of domain seperation here too?
+	return h.Sum(nil)
```

Comment:
In what circumstances would a hash of the root with its namespaces prefix be necessary? Sending the actual root with the prefixes doesn't take up much more bandwidth.

Reply:
@adlerjohn argued that the library should rather have a method to return a compact commitment to of the prefixed root (e.g. via `hash(minNs||maxNs||actualRoot)`).

Reply:
removed in https://github.com/lazyledger/nmt/pull/3/commits/fd6c90fa94c346a153d7f9537a4cb64a70273670 for now as the caller can handle this as required